File: 03-array/17-maximum-subarray-sum/17-maximum-subarray-sum.py
from typing import List

# * The idea is max of i is the maximum of (i-1 value + i th value) or i th value
# * Time: O(n) Space: O(1)
def maximum_subarray_sum(nums: List[int]) -> int:
  max_sub_arr_sum, curr_max_sub_arr_sum = nums[0], nums[0]
  n = len(nums)
  for i in range(1, n):
    curr_max_sub_arr_sum = max(curr_max_sub_arr_sum + nums[i], nums[i])
    max_sub_arr_sum = max(max_sub_arr_sum, curr_max_sub_arr_sum)
  return max_sub_arr_sum


print(maximum_subarray_sum([-5, 1, -2, 3, -1, 2, -2]))

File: 03-array/17-maximum-subarray-sum/17-maximum-subarray-sum.ts
const max_subarray_sum = (nums: number[]): number => {
  let [curr_max_sub_arr_sum, max_sub_arr_sum] = [nums[0], nums[0]]
  for (let i = 1; i < nums.length; i++) {
    if (curr_max_sub_arr_sum + nums[i] > nums[i]) {
      curr_max_sub_arr_sum += nums[i]
    } else {
      curr_max_sub_arr_sum = nums[i]
    }
    max_sub_arr_sum = Math.max(curr_max_sub_arr_sum, max_sub_arr_sum)
  }
  return max_sub_arr_sum
}

console.log(max_subarray_sum([-5, 1, -2, 3, -1, 2, -2]))
